ipfsconn/ipfshttp/ipfshttp.go

```diff
@@ -433,31 +446,30 @@ func (ipfs *Connector) Unpin(ctx context.Context, hash cid.Cid) error {
 	ctx, span := trace.StartSpan(ctx, "ipfsconn/ipfshttp/Unpin")
 	defer span.End()
 
-	pinStatus, err := ipfs.PinLsCid(ctx, hash)
-	if err != nil {
-		return err
+	if ipfs.config.UnpinDisable {
+		return errors.New("ipfs unpinning is disallowed by configuration on this peer")
 	}
 
-	if pinStatus.IsPinned(-1) {
-		if ipfs.config.UnpinDisable {
-			return errors.New("ipfs unpinning is disallowed by configuration on this peer")
-		}
+	defer ipfs.updateInformerMetric(ctx)
 
-		defer ipfs.updateInformerMetric(ctx)
-		path := fmt.Sprintf("pin/rm?arg=%s", hash)
+	path := fmt.Sprintf("pin/rm?arg=%s", hash)
 
-		ctx, cancel := context.WithTimeout(ctx, ipfs.config.UnpinTimeout)
-		defer cancel()
+	ctx, cancel := context.WithTimeout(ctx, ipfs.config.UnpinTimeout)
+	defer cancel()
 
-		_, err := ipfs.postCtx(ctx, path, "", nil)
-		if err != nil {
+	// We will call unpin in any case, if the CID is not pinned,
+	// then we ignore the error (although this is a bit flaky).
+	_, err := ipfs.postCtx(ctx, path, "", nil)
+	if err != nil {
+		ipfsErr, ok := err.(ipfsError)
+		if !ok || ipfsErr.Message != pinner.ErrNotPinned.Error() {
 			return err
 		}
-		logger.Info("IPFS Unpin request succeeded:", hash)
-		stats.Record(ctx, observations.Pins.M(-1))
+		logger.Debug("IPFS object is already unpinned: ", hash)
 	}
 
-	logger.Debug("IPFS object is already unpinned: ", hash)
+	logger.Info("IPFS Unpin request succeeded:", hash)
+	stats.Record(ctx, observations.Pins.M(-1))
```

Comment:
For unpins, we stopped checking if the item is pinned with PinLsCid. This means 2 calls for every unpin, makes IPFS load the whole recursive pinset list to check. We just trigger unpin and check the error message, succeeding if it's an ErrNotPinned. It is not awesome that we have to parse an error message, but IPFS gives little change (Error code is set to 0 and http status code is 500, so it's indistinguisable from other errors).